A42-xds-ring-hash-lb-policy.md

```diff
@@ -257,6 +257,10 @@ hard-coded LB policy config fields in the CDS response:
 - For `RING_HASH`, the policy name will be `"RING_HASH"`, and the config will
   be the one for the `ring_hash_experimental` LB Policy described below.
 
+If the `xds_cluster_resolver` config is generated from an aggregate cluster
```

Comment:
Hmm... Now that I'm looking at the text here, why is this necessary to say in A42?  The aggregate cluster functionality was actually introduced in A37, and the code described in this text was added to the CDS policy as part of implementing A37.  Nothing in A42 actually changes this.

I think this text should instead go in this section in A37:

https://github.com/grpc/proposal/blob/master/A37-xds-aggregate-and-logical-dns-clusters.md#cds-lb-policy-support-for-aggregate-and-logical-dns-clusters

Reply:
A37 does not describe any mechanism at all for populating the `locality_picking_policy` and `endpoint_picking_policy` fields of the `xds_cluster_resolver` config, so it doesn't make sense to me to add a clarification there of how those fields should be populated in one specific scenario. A42 is the first gRFC that describes such a mechanism, so it makes the most sense to me to clarify its behavior there.

Reply:
Also, I think the "one specific scenario" you mention is a reference to aggregate clusters, and that's exactly what A37 introduces, so I think it's totally on-topic to cover it there.

Reply:
"Mechanism" may not have been exactly the right word here. A37 does not describe what fields in the CDS update should be used to determine the values of the `locality_picking_policy` and `endpoint_picking_policy` fields.

Reply:
If we add a statement in A37 that says "In the aggregate cluster case, the values of the `locality_picking_policy` and `endpoint_picking_policy` fields should be determined from the root cluster update," then an obvious next question most readers will have is "*How* should the values of those fields be determined from the root cluster update?" Since A37 does not answer this question, this change would increase confusion rather than resolve it.

Reply:
Okay, I had to go digging back through some old PRs to remind myself of the history here, but I now see what happened.  The aggregate cluster functionality was implemented in C-core in https://github.com/grpc/grpc/pull/25078.  At the time, we didn't actually support any LB policies except for ROUND_ROBIN, so the CDS validation code rejected any resource with any other policy, and there actually wasn't any field in the CDS update indicating the LB policy, because there was no possible value other than ROUND_ROBIN and therefore no reason to specify it.  The CDS policy was simply hard-coding the values in the `xds_cluster_resolver` config: it set `locality_picking_policy` to `weighted_target` and `endpoint_picking_policy` to `round_robin`.  (We originally added those two fields in the `eds_experimental` policy that was created in the internal design go/grpc-xds-policy-refactoring, which was then split into `xds_cluster_resolver` and `xds_cluster_impl` in A37.  One of the goals of go/grpc-xds-policy-refactoring was to lay the groundwork for hierarchical LB policies, which is why these fields were added there.)

So you're right in the sense that the question of which cluster's LB policy config to use in an aggregate cluster was not relevant until A42, since that's the first time we had a choice of LB policies.  But at the same time, I still think that the fact that the LB policy comes from the aggregate cluster instead of the underlying clusters is a key part of the definition of an aggregate cluster and therefore needs to be specified in A37.  So, I propose the following:

- In A37, add text saying something like this: "In an aggregate cluster, the LB policy config comes from the aggregate cluster, not the underlying clusters.  At present, we do not support any LB policy except ROUND_ROBIN, so this decision is moot, and the CDS policy will hard-code the `locality_picking_policy` field to `weighted_target` and the `endpoint_picking_policy` field to `round_robin`.  However, in the future, when we add support for other LB policies, the CDS policy will be changed to get the LB policy config from the aggregate cluster."
- Here in A42, say something like this: "As mentioned in gRFC A37, the LB policy config for an aggregate cluster comes from the aggregate cluster, not the underlying clusters.  The CDS policy will select that config when constructing the `xds_lb_policy` field for the `xds_cluster_resolver` policy."

Reply:
My interpretation of A37 when I implemented it was that the `locality_picking_policy` and `endpoint_picking_policy` could be empty (since the documentation describes them as "optional") and that `xds_cluster_resolver` policy should know that the defaults are `weighted_target` and `round_robin` respectively, and fall back to those defaults if the fields are empty. So my CDS policy just left those fields empty and let the child policy sort it out.

Reply:
Sure, that's fine.  I don't think that affects the substance of what we're talking about here.  Regardless of whether the CDS policy leaves the fields unset and lets them default or explicitly sets them to the same defaults, the overall structure is still the same in terms of where decisions are being made.

Reply:
I made the change to A37, and updated the wording of the others. I kept the mention of this point in A52, because A52 is making some changes to the CDS update and the logic for converting it to an `xds_cluster_resolver` config, and I think it's worth mentioning explicitly that this part of that logic is not changing.